fix: Handle error when offscreen document already exists

Gudahtt opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

A `try..catch` block has been added to catch errors about the Offscreen document already existing. This circumstance is not a bug, it's expected to happen sometimes due to race conditions between `hasDocument` and the creation step. The code in question is just meant to create the document if it doesn't exist, so it's OK if we get this error.

The `hasDocument` call has been removed as well, since it's now redundant. This function wasn't safe to call anyway; it's an undocumented private function.

See here for more information:
* https://stackoverflow.com/a/7725839
* https://groups.google.com/a/chromium.org/g/chromium-extensions/c/D5Jg2ukyvUc4

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/25138?quickstart=1)

## **Related issues**

Fixes #25118

## **Manual testing steps**

I do not know how to test this unfortunately.

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Gudahtt committed:
fix: Handle error when offscreen document already exists

A `try..catch` block has been added to catch errors about the Offscreen
document already existing. This circumstance is not a bug, it's
expected to happen sometimes due to race conditions between
`hasDocument` and the creation step. The code in question is just meant
to create the document if it doesn't exist, so it's OK if we get this
error.

The `hasDocument` call has been removed as well, since it's now
redundant. This function wasn't safe to call anyway; it's an
undocumented private function.

See here for more information:
* https://stackoverflow.com/a/7725839
* https://groups.google.com/a/chromium.org/g/chromium-extensions/c/D5Jg2ukyvUc4

Fixes #25118

Gudahtt commented:
I can't write unit tests for this file directly, not easily at least, because it has side-effects. I could move this function (and maybe some others) into a separate module and write unit tests for them there. Probably we should do this at some point. But there is minimal value in doing so because the logic here is so simple. Really it's the chrome extension API behavior that would be valuable to test; if we're mocking that, we're not really validating this fix.

That means a valuable test would have to be an e2e or manual test. But I don't know how to trigger the race condition where we attempt to create the offscreen document when it already exists.

danjm commented:
don't merge until https://github.com/MetaMask/metamask-extension/pull/25142 is merged

Gudahtt commented:
Still investigating the remaining e2e test error. It's a pre-existing flaky test, but something about this PR appears to make it fail more often (possibly every time)

Gudahtt commented:
The test failures on this PR were resolved by #25164

Gudahtt committed:
Update error handling to avoid crashing

We now capture unrecognized errors and report them to Sentry, rather
than throwing them. This ensures that any new failures would not brick
the wallet.

Gudahtt commented on app/scripts/offscreen.js:
```diff
@@ -9,29 +10,51 @@ import { OffscreenCommunicationTarget } from '../../shared/constants/offscreen-c
+      // Report unrecongized errors without halting wallet initialization
+      // Failures to create the offscreen document does not compromise wallet data integrity or
+      // core functionality, it's just needed for specific features.
+      captureException(error);
```
Just updated this to catch all platform errors here (as suggested by @FrederikBolding on Slack) to avoid bricking the wallet if something else goes wrong here. We're still capturing the error in Sentry, and this is a non-essential step, so this should be a safe way to approach this for now.

ffb2c3a3689b20ee7f543dfb6be8658745297e8e
